### PR TITLE
Floating-point rounding instructions

### PR DESCRIPTION
**Pull request by Maratyszcza**
Introduction
========

Floating-point round-to-integer is a widely used operation, available in many software and hardware specifications:
- As `f32.nearest`/`f32.trunc`/`f32.ceil`/`f32.floor`/`f64.nearest`/`f64.trunc`/`f64.ceil`/`f64.floor` scalar instruction in WebAssembly
- As `rint`/`nearbyint`/`trunc`/`ceil`/`floor` functions in C and C++
- As `ROUNDPS` and `ROUNDPD` instructions in SSE4.1
- As `VRINTN`/`VRINTZ`/`VRINTP`/`VRINTM` instructions in ARMv8 AArch32
- As `FRINTN`/`FRINTZ`/`FRINTP`/`FRINTM` instructions in AArch64

These PR introduce the rounding instructions in WebAssembly SIMD.

New instructions
===========

- Round to nearest integer, ties to even: `f32x4.nearest`/`f64x2.nearest`
- Round to integer towards zero (truncate to integer): `f32x4.trunc`/`f64x2.trunc`
- Round to integer above (ceiling): `f32x4.ceil`/`f64x2.ceil`
- Round to integer below (floor): `f32x4.floor`/`f64x2.floor`

The instructions match the scalar WebAssembly analogs both in names and in semantics.

Mapping to Common Instruction Sets
=============================

This section illustrates how the new WebAssembly instructions can be lowered on common instruction sets. However, these patterns are provided only for convenience, compliant WebAssembly implementations do not have to follow the same code generation patterns.

x86/x86-64 processors with AVX instruction set
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` is lowered to `VROUNDPS xmm_y, xmm_x, 0x08`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` is lowered to `VROUNDPS xmm_y, xmm_x, 0x0B`
- **f32x4.ceil**
  - `y = f32x4.ceil(x)` is lowered to `VROUNDPS xmm_y, xmm_x, 0x0A`
- **f32x4.floor**
  - `y = f32x4.floor(x)` is lowered to `VROUNDPS xmm_y, xmm_x, 0x09`
- **f64x2.nearest**
  - `y = f64x2.nearest(x)` is lowered to `VROUNDPD xmm_y, xmm_x, 0x08`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` is lowered to `VROUNDPD xmm_y, xmm_x, 0x0B`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` is lowered to `VROUNDPD xmm_y, xmm_x, 0x0A`
- **f64x2.floor**
  - `y = f64x2.floor(x)` is lowered to `VROUNDPD xmm_y, xmm_x, 0x09`

x86/x86-64 processors with SSE4.1 instruction set
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` is lowered to `ROUNDPS xmm_y, xmm_x, 0x08`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` is lowered to `ROUNDPS xmm_y, xmm_x, 0x0B`
- **f32x4.ceil**
  - `y = f32x4.ceil(x)` is lowered to `ROUNDPS xmm_y, xmm_x, 0x0A`
- **f32x4.floor**
  - `y = f32x4.floor(x)` is lowered to `ROUNDPS xmm_y, xmm_x, 0x09`
- **f64x2.nearest**
  - `y = f64x2.nearest(x)` is lowered to `ROUNDPD xmm_y, xmm_x, 0x08`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` is lowered to `ROUNDPD xmm_y, xmm_x, 0x0B`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` is lowered to `ROUNDPD xmm_y, xmm_x, 0x0A`
- **f64x2.floor**
  - `y = f64x2.floor(x)` is lowered to `ROUNDPD xmm_y, xmm_x, 0x09`

x86/x86-64 processors with SSE2 instruction set
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVDQA   xmm_tmp0, wasm_splat_u32(0x80000000)`
    - `CVTPS2DQ xmm_y, xmm_x`
    - `CVTDQ2PS xmm_tmp1, xmm_y`
    - `PCMPEQD  xmm_y, xmm_tmp0`
    - `POR      xmm_y, xmm_tmp0`
    - `ADDPS    xmm_tmp0, xmm_x`
    - `ANDPS    xmm_tmp0, xmm_y`
    - `ANDNPS   xmm_y, xmm_tmp1`
    - `ORPS     xmm_y, xmm_tmp0`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVDQA    xmm_tmp0, wasm_splat_u32(0x80000000)`
    - `CVTTPS2DQ xmm_y, xmm_x`
    - `CVTDQ2PS  xmm_tmp1, xmm_y`
    - `PCMPEQD   xmm_y, xmm_tmp0`
    - `POR       xmm_y, xmm_tmp0`
    - `ADDPS     xmm_tmp0, xmm_x`
    - `ANDPS     xmm_tmp0, xmm_y`
    - `ANDNPS    xmm_y, xmm_tmp1`
    - `ORPS      xmm_y, xmm_tmp0`
- **f32x4.ceil**
  - `x = f32x4.ceil(x)` is lowered to:
    - `CVTTPS2DQ xmm_tmp0, xmm_x`
    - `MOVDQA    xmm_tmp1, wasm_splat_u32(0x80000000)`
    - `CVTDQ2PS  xmm_tmp2, xmm_tmp0`
    - `PCMPEQD   xmm_tmp0, xmm_tmp1`
    - `POR       xmm_tmp0, xmm_tmp1`
    - `MOVDQA    xmm_tmp3, xmm_tmp0`
    - `ANDPS     xmm_tmp3, xmm_x`
    - `ANDNPS    xmm_tmp0, xmm_tmp2`
    - `ORPS      xmm_tmp0, xmm_tmp3`
    - `CMPLEPS   xmm_x, xmm_tmp0`
    - `ORPS      xmm_x, xmm_tmp1`
    - `MOVAPS    xmm_tmp2, xmm_x`
    - `ANDPS     xmm_tmp2, xmm_tmp0`
    - `ADDPS     xmm_tmp0, wasm_splat_f32(1.0f)`
    - `ANDNPS    xmm_x, xmm_tmp0`
    - `ORPS      xmm_x, xmm_tmp2`
- **f32x4.floor**
  - `y = f32x4.floor(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVDQA    xmm_tmp0, wasm_splat_u32(0x80000000)`
    - `CVTTPS2DQ xmm_y, xmm_x`
    - `CVTDQ2PS  xmm_tmp1, xmm_y`
    - `PCMPEQD   xmm_y, xmm_tmp0`
    - `POR       xmm_y, xmm_tmp0`
    - `MOVAPS    xmm_tmp0, xmm_y`
    - `ANDPS     xmm_tmp0, xmm_x`
    - `ANDNPS    xmm_y, xmm_tmp1`
    - `MOVAPS    xmm_tmp1, xmm_x`
    - `ORPS      xmm_y, xmm_tmp0`
    - `CMPLTPS   xmm_tmp1, xmm_y`
    - `ANDPS     xmm_tmp1, wasm_splat_f32(1.0f)`
    - `SUBPS     xmm_y, xmm_tmp1`
- **f64x2.nearest**
  - `y = f64x2.nearest(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVAPS  xmm_tmp0, wasm_splat_u64(0x7FFFFFFFFFFFFFFF)`
    - `MOVAPS  xmm_y, xmm_x`
    - `MOVAPS  xmm_tmp1, wasm_splat_f64(0x1.0p+52)`
    - `MOVAPS  xmm_tmp2, xmm_tmp0`
    - `ANDPS   xmm_y, xmm_tmp1`
    - `CMPLEPD xmm_tmp2, xmm_y`
    - `ADDPD   xmm_y, xmm_tmp0`
    - `SUBPD   xmm_y, xmm_tmp0`
    - `ANDNPS  xmm_tmp2, xmm_tmp1`
    - `MOVAPS  xmm_tmp1, xmm_tmp2`
    - `ANDNPS  xmm_tmp1, xmm_x`
    - `ANDPS   xmm_y, xmm_tmp2`
    - `ORPS    xmm_y, xmm_tmp1`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVAPS   xmm_y, wasm_splat_u64(0x7FFFFFFFFFFFFFFF)`
    - `MOVAPS   xmm_tmp0, wasm_splat_f64(0x1.0p+52)`
    - `MOVAPS   xmm_tmp1, xmm_x`
    - `ANDPS    xmm_tmp1, xmm_y`
    - `MOVAPS   xmm_tmp2, xmm_tmp0`
    - `CMPNLEPD xmm_tmp2, xmm_tmp1`
    - `ANDPS    xmm_y, xmm_tmp2`
    - `MOVAPS   xmm_tmp2, xmm_tmp1`
    - `ADDPD    xmm_tmp2, xmm_tmp0`
    - `SUBPD    xmm_tmp2, xmm_tmp0`
    - `CMPLTPD  xmm_tmp1, xmm_tmp2`
    - `ANDPS    xmm_tmp1, wasm_splat_f64(1.0)`
    - `SUBPD    xmm_tmp2, xmm_tmp1`
    - `ANDPS    xmm_tmp2, xmm_y`
    - `ANDNPS   xmm_y, xmm_x`
    - `ORPS     xmm_y, xmm_tmp2`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVAPS   xmm_tmp0, wasm_splat_u64(0x7FFFFFFFFFFFFFFF)`
    - `MOVAPS   xmm_y, xmm_x`
    - `MOVAPS   xmm_tmp1, wasm_splat_f64(0x1.0p+52)`
    - `ANDPS    xmm_y, xmm_tmp0`
    - `MOVAPS   xmm_tmp2, xmm_tmp1`
    - `CMPNLEPD xmm_tmp2, xmm_y`
    - `ADDPD    xmm_y, xmm_tmp1`
    - `ANDPS    xmm_tmp2, xmm_tmp0`
    - `SUBPD    xmm_y, xmm_tmp1`
    - `ANDPS    xmm_y, xmm_tmp2`
    - `ANDNPS   xmm_tmp2, xmm_x`
    - `ORPS     xmm_tmp2, xmm_y`
    - `MOVAPS   xmm_y, xmm_tmp2`
    - `MOVAPS   xmm_tmp1, xmm_tmp2`
    - `CMPLTPD  xmm_y, xmm_x`
    - `ADDPD    xmm_tmp1, wasm_splat_f64(1.0)`
    - `ANDPS    xmm_y, xmm_tmp0`
    - `ANDPS    xmm_tmp1, xmm_y`
    - `ANDNPS   xmm_y, xmm_tmp2`
    - `ORPS     xmm_y, xmm_tmp1`
- **f64x2.floor**
  - `y = f64x2.floor(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVAPS   xmm_tmp0, wasm_splat_u64(0x7FFFFFFFFFFFFFFF)`
    - `MOVAPS   xmm_tmp1, xmm_x`
    - `MOVAPS   xmm_tmp2, wasm_splat_f64(0x1.0p+52)`
    - `ANDPS    xmm_tmp1, xmm_tmp0`
    - `MOVAPS   xmm_y, xmm_tmp2`
    - `CMPNLEPD xmm_y, xmm_tmp1`
    - `ANDPS    xmm_y, xmm_tmp0`
    - `ADDPD    xmm_tmp1, xmm_tmp2`
    - `SUBPD    xmm_tmp1, xmm_tmp2`
    - `ANDPS    xmm_tmp1, xmm_y`
    - `ANDNPS   xmm_y, xmm_x`
    - `MOVAPS   xmm_tmp0, xmm_x`
    - `ORPS     xmm_y, xmm_tmp1`
    - `CMPLTPD  xmm_tmp0, xmm_y`
    - `ANDPS    xmm_tmp0, wasm_splat_f64(1.0)`
    - `SUBPD    xmm_y, xmm_tmp0`

ARM64 processors
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` is lowered to `FRINTN Vy.4S, Vx.4S`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` is lowered to `FRINTZ Vy.4S, Vx.4S`
- **f32x4.ceil**
  - `y = f32x4.ceil(x)` is lowered to `FRINTP Vy.4S, Vx.4S`
- **f32x4.floor**
  - `y = f32x4.floor(x)` is lowered to `FRINTM Vy.4S, Vx.4S`
- **f64x2.nearest**
  - `y = f64x2.nearest(x)` is lowered to `FRINTN Vy.2D, Vx.2D`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` is lowered to `FRINTZ Vy.2D, Vx.2D`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` is lowered to `FRINTP Vy.2D, Vx.2D`
- **f64x2.floor**
  - `y = f64x2.floor(x)` is lowered to `FRINTM Vy.2D, Vx.2D`

ARM processors with ARMv8 (32-bit) instruction set
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` is lowered to `VRINTN.F32 Qy, Qx`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` is lowered to `VRINTZ.F32 Qy, Qx`
- **f32x4.ceil**
  - `y = f32x4.ceil(x)` is lowered to `VRINTP.F32 Qy, Qx`
- **f32x4.floor**
  - `y = f32x4.floor(x)` is lowered to `VRINTM.F32 Qy, Qx`
- **f64x2.nearest**
  - `y = f64x2.nearest(x)` is lowered to `VRINTN.F64 Dy_lo, Dx_lo` + `VRINTN.F64 Dy_hi, Dx_hi`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` is lowered to `VRINTZ.F64 Dy_lo, Dx_lo` + `VRINTZ.F64 Dy_hi, Dx_hi`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` is lowered to `VRINTP.F64 Dy_lo, Dx_lo` + `VRINTP.F64 Dy_hi, Dx_hi`
- **f64x2.floor**
  - `y = f64x2.floor(x)` is lowered to `VRINTM.F64 Dy_lo, Dx_lo` + `VRINTM.F64 Dy_hi, Dx_hi`

ARM processors with ARMv7 (32-bit) instruction set
-----------------------------------------------

- **f32x4.nearest**
  - `y = f32x4.nearest(x)` (`y` is **NOT** `x`) is lowered to:
    - `VMOV.I32  Qtmp0, 0x4B000000`
    - `VABS.F32  Qtmp1, Qx`
    - `VACGT.F32 Qy,  Qx,  Qtmp0`
    - `VADD.F32  Qtmp1, Qtmp1, Qtmp0`
    - `VORR.I32  Qy,  0x80000000`
    - `VSUB.F32  Qtmp1, Qtmp1, Qtmp0`
    - `VBSL      Qy,  Qx,  Qtmp1`
- **f32x4.trunc**
  - `y = f32x4.trunc(x)` (`y` is **NOT** `x`) is lowered to:
    - `VCVT.S32.F32 Qtmp0, Qx`
    - `VMOV.I32     Qtmp1, 0x4B000000`
    - `VACGT.F32    Qy, Qtmp1, Qx`
    - `VCVT.F32.S32 Qtmp0, Qtmp0`
    - `VBIC.I32     Qy, 0x80000000`
    - `VBSL         Qy, Qtmp0, Qx`
- **f32x4.ceil**
  - `y = f32x4.ceil(x)` (`y` is **NOT** `x`) is lowered to:
    - `VCVT.S32.F32 Qtmp0, Qx`
    - `VMOV.I32     Qtmp1, 0x4B000000`
    - `VACGT.F32    Qtmp1, Qtmp1, Qx`
    - `VCVT.F32.S32 Qtmp0, Qtmp0`
    - `VBIC.I32     Qtmp1, 0x80000000`
    - `VBSL         Qtmp1, Qtmp0, Qx`
    - `VMOV.F32     Qtmp0, 0x3F800000`
    - `VCGE.F32     Qy, Qtmp1, Qx`
    - `VADD.F32     Qtmp0, Qtmp1, Qtmp0`
    - `VORR.I32     Qy, 0x80000000`
    - `VBSL         Qy, Qtmp1, Qtmp0`
- **f32x4.floor**
  - `y = f32x4.floor(x)` (`y` is **NOT** `x`) is lowered to:
    - `VCVT.S32.F32 Qtmp0, Qx`
    - `VMOV.I32     Qtmp1, 0x4B000000`
    - `VACGT.F32    Qy, Qtmp1, Qx`
    - `VCVT.F32.S32 Qtmp0, Qtmp0`
    - `VBIC.I32     Qy, 0x80000000`
    - `VBSL         Qy, Qtmp0, Qx`
    - `VMOV.F32     Qtmp1, 0x3F800000`
    - `VCGT.F32     Qtmp0, Qy, Qx`
    - `VAND         Qtmp0, Qtmp0, Qtmp1`
    - `VSUB.F32     Qy, Qy, Qtmp0`
- **f64x2.round**
  - `y = f64x2.round(x)` (`y` is **NOT** `x`) is lowered to:
    - `VABS.F64 Dy_lo, Dx_lo`
    - `VABS.F64 Dy_hi, Dx_hi`
    - `VLDR     Dtmp0, 0x1.0p+52`
    - `VSUB.F64 Dtmp1_lo, Dtmp0, Dy_lo`
    - `VSUB.F64 Dtmp1_hi, Dtmp0, Dy_hi`
    - `VADD.F64 Dtmp2_lo, Dy_lo, Dtmp0`
    - `VADD.F64 Dtmp2_hi, Dy_hi, Dtmp0`
    - `VEOR     Qy, Qx, Qy`
    - `VSHR.S64 Qtmp1, Qtmp1, 63`
    - `VSUB.F64 Dtmp2_lo, Dtmp2_lo, Dtmp0`
    - `VSUB.F64 Dtmp2_hi, Dtmp2_hi, Dtmp0`
    - `VORR     Qy, Qy, Qtmp1`
    - `VBSL     Qy, Qx, Qtmp2`
- **f64x2.trunc**
  - `y = f64x2.trunc(x)` (`y` is **NOT** `x`) is lowered to:
    - `VLDR     Dtmp0, 0x1.0p+52`
    - `VABS.F64 Qy_lo, Dx_lo`
    - `VABS.F64 Qy_hi, Dx_hi`
    - `VADD.F64 Dtmp1_lo, Qy_lo, Dtmp0`
    - `VADD.F64 Dtmp1_hi, Qy_hi, Dtmp0`
    - `VSUB.F64 Dtmp2_lo, Dtmp0, Qy_lo`
    - `VSUB.F64 Dtmp2_hi, d9, Qy_hi`
    - `VEOR     Qtmp3, Qy, Qx`
    - `VSUB.F64 Dtmp1_lo, Dtmp1_lo, Dtmp0`
    - `VSUB.F64 Dtmp1_hi, Dtmp1_hi, d9`
    - `VLDR     Dtmp0, 1.0`
    - `VSHR.S64 Qtmp2, Qtmp2, 63`
    - `VORR     Qtmp3, Qtmp3, Qtmp2`
    - `VSUB.I64 Qy, Qy, Qtmp1`
    - `VSHR.S64 Qy, Qy, 63`
    - `VAND     Qy_lo, Qy_lo, Dtmp0`
    - `VAND     Qy_hi, Qy_hi, Dtmp0`
    - `VSUB.F64 Qy_lo, Dtmp1_lo, Qy`
    - `VSUB.F64 Qy_hi, Dtmp1_hi, Qx`
    - `VBIT     Qy, Qx, Qtmp3`
- **f64x2.ceil**
  - `y = f64x2.ceil(x)` (`y` is **NOT** `x`) is lowered to:
    - `VLDR     Dtmp0, 0x1.0p+52`
    - `VABS.F64 Dtmp1_lo, Dx_lo`
    - `VABS.F64 Dtmp1_hi, Dx_hi`
    - `VSUB.F64 Dtmp2_lo, Dtmp0, Dtmp1_lo`
    - `VSUB.F64 Dtmp2_hi, Dtmp0, Dtmp1_hi`
    - `VADD.F64 Dtmp3_lo, Dtmp1_lo, Dtmp0`
    - `VADD.F64 Dtmp3_hi, Dtmp1_hi, Dtmp0`
    - `VEOR     Qtmp1, Qtmp1, Qx`
    - `VSHR.S64 Qtmp2, Qtmp2, 63`
    - `VSUB.F64 Dtmp3_lo, Dtmp3_lo, Dtmp0`
    - `VSUB.F64 Dtmp3_hi, Dtmp3_hi, Dtmp0`
    - `VLDR     Dtmp0, 1.0`
    - `VORR     Qtmp2, Qtmp2, Qtmp1`
    - `VBSL     Qtmp2, Qx, Qtmp3`
    - `VSUB.F64 Dy_lo, Dtmp2_lo, Dx_lo`
    - `VSUB.F64 Dy_hi, Dtmp2_hi, Dx_hi`
    - `VADD.F64 Dtmp3_lo, Dtmp2_lo, Dtmp0`
    - `VADD.F64 Dtmp3_hi, Dtmp2_hi, Dtmp0`
    - `VSHR.S64 Qy, Qy, 63`
    - `VBIC     Qy, Qy, Qtmp1`
    - `VBSL     Qy, Qtmp3, Qtmp2`
- **f64x2.floor**
  - `y = f64x2.floor(x)` (`y` iD **NOT** `x`) iD lowereQ to:
    - `VLDR     Dtmp0, 0x1.0p+52`
    - `VABS.F64 Dy_lo, Dx_lo`
    - `VABS.F64 Dy_hi, Dx_hi`
    - `VADD.F64 Dtmp1_lo, Dy_lo, Dtmp0`
    - `VADD.F64 Dtmp1_hi, Dy_hi, Dtmp0`
    - `VSUB.I64 Dtmp2_lo, Dtmp0, Dy_lo`
    - `VSUB.I64 Dtmp2_hi, Dtmp0, Dy_hi`
    - `VEOR     Qy, Qy, Qx`
    - `VSUB.F64 Dtmp1_lo, Dtmp1_lo, Dtmp0`
    - `VSUB.F64 Dtmp1_hi, Dtmp1_hi, Dtmp0`
    - `VLDR     Dtmp0, 1.0`
    - `VSHR.S64 Qtmp2, Qtmp2, 63`
    - `VORR     Qy, Qy, Qtmp2`
    - `VBSL     Qy, Qx, Qtmp1`
    - `VSUB.F64 Dx_lo, Dx_lo, Dy_lo`
    - `VSUB.F64 Dx_hi, Dx_hi, Dy_hi`
    - `VSHR.S64 Qtmp2, Qx, 63`
    - `VAND     Dtmp2_lo, Dtmp2_lo, Dtmp0`
    - `VAND     Dtmp2_hi, Dtmp2_hi, Dtmp0`
    - `VSUB.F64 Dy_lo, Dy_lo, Dtmp2_lo`
    - `VSUB.F64 Dy_hi, Dy_hi, Dtmp2_hi`

### CODE REVIEW

**Comment by tlively** [comment]
Yikes, the new numbering only has room for one rounding instruction. We'll have to figure out what to do about that in the long term. Meanwhile, @dtig and @ngzhian do you have a preference about which opcodes to use to prototype this?

**Comment by ngzhian** [comment]
No preferences for prototyping, we can probably squeeze them into
0xdc-0xdf
0xe2, 0xee, 0xf8, 0xf9
for now.

**Comment by dtig** [comment]
No strong preferences either, it's somewhat awkward, but we could also do something in the range of 0xc2- 0xca if contiguous opcodes make this simpler, because I don't see the 64x2 AnyTrue/AllTrue and the widen/narrowing instructions to be relevant for 64x2 operations going forward. 

If we do have to spill over, it's not terrible but we can make that call when we decide to move past prototyping.

**Comment by richgel999** [comment]
These instructions aren't optional IMO. They're fundamental operations. Having to emulate them will be quite painful for many SIMD/SPMD kernels and vectorized math functions. 

I have a Perlin noise kernel that computes 24 floors per output pixels:
https://t.co/u9w35T6oTq?amp=1

In another example, I have vectorized approximate math library. It can compute vectorized tan, sin, cos, log, exp, etc. It uses floor and round for range reduction:
https://t.co/3JlYyZ2oMI?amp=1

Without efficient round/floor/trunc, WebAssembly SIMD will be in the same position SSE2 is relative to SSE4.1. When we execute kernels on SSE2, we commonly get a 15-20% reduction in performance due to having to emulate round/floor/trunc on some kernels, or if they call sin/cos/tan/etc. These are very important operations.

**Comment by zeux** [comment]
Worth noting is that the common way to emulate round/floor/trunc includes conversions back & forth to integers (obviously this is application-dependent as it assumes a specific range and is typically non-IEEE compliant for some operations); however, due to #173 this workaround is going to be slow.

If the inputs are known to be within a 23-bit integer range or thereabouts, floating point addition can be abused to round, and it's probably possible to implement floor etc. in a similar fashion but that route doesn't seems like one we would want to recommend.

**Comment by Marc-B-Reynolds** [comment]
> If the inputs are known to be within a 23-bit integer range or thereabouts, floating point addition can be abused to round, and it's probably possible to implement floor etc. in a similar fashion but that route doesn't seems like one we would want to recommend.

Worth nothing this this stops working if FP rules are relaxed:  `(x+K)-K` x-formed to `x`

**Comment by ngzhian** [comment]
@Maratyszcza any suggestions for ARM v7 instruction sequence? It will probably look a lot like the x86 SSE2 one?

**Comment by Maratyszcza** [comment]
Updated opcodes post-renumbering, put into 0xd8-0xdf range

**Comment by Maratyszcza** [comment]
Mapping to SSE2 is finished. @ngzhian ARMv7 NEON is quite different, because of its unique features:
- Compare absolute values instruction
- Single-instruction bitwise selection (`VBSL`/`VBIT`/`VBIF`)
- Bitwise OR and bitwise AND instructions with immediate values

**Comment by Maratyszcza** [comment]
Added ARMv7 NEON mapping for `f32` instructions

**Comment by ngzhian** [comment]
There's some magic going on there. Thanks Marat!

**Comment by Maratyszcza** [comment]
All instructions mappings are finished, and PR is ready for review

File: proposals/simd/BinarySIMD.md [hunk 1]
```diff
@@ -183,6 +183,14 @@ The `v8x16.shuffle` instruction has 16 bytes after `simdop`.
...
 | `i64x2.sub`                |    `0xd1`| -                  |
 | `i64x2.mul`                |    `0xd5`| -                  |
+| `f32x4.nearest`            |    `0xd8`| -                  |
+| `f32x4.trunc`              |    `0xd9`| -                  |
+| `f32x4.ceil`               |    `0xda`| -                  |
+| `f32x4.floor`              |    `0xdb`| -                  |
+| `f64x2.nearest`            |    `0xdc`| -                  |
+| `f64x2.trunc`              |    `0xdd`| -                  |
+| `f64x2.ceil`               |    `0xde`| -                  |
+| `f64x2.floor`              |    `0xdf`| -                  |
```
Comment (tlively) [inline proposals/simd/BinarySIMD.md:193]:
```suggestion
| `f32x4.ceil`               |    `0xd8`| -                  |
| `f32x4.floor`              |    `0xd9`| -                  |
| `f32x4.trunc`              |    `0xda`| -                  |
| `f32x4.nearest`            |    `0xdb`| -                  |
| `f64x2.ceil`               |    `0xdc`| -                  |
| `f64x2.floor`              |    `0xdd`| -                  |
| `f64x2.trunc`              |    `0xde`| -                  |
| `f64x2.nearest`            |    `0xdf`| -                  |
```

File: proposals/simd/ImplementationStatus.md [hunk 2]
```diff
@@ -165,6 +165,10 @@
 | `f32x4.div`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
 | `f32x4.min`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
 | `f32x4.max`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
+| `f32x4.nearest`            |                           |                    |                    |                    |                    |
+| `f32x4.trunc`              |                           |                    |                    |                    |                    |
+| `f32x4.ceil`               |                           |                    |                    |                    |                    |
+| `f32x4.floor`              |                           |                    |                    |                    |                    |
```
Comment (tlively) [inline proposals/simd/ImplementationStatus.md:171]:
```suggestion
| `f32x4.ceil`               |                           |                    |                    |                    |                    |
| `f32x4.floor`              |                           |                    |                    |                    |                    |
| `f32x4.trunc`              |                           |                    |                    |                    |                    |
| `f32x4.nearest`            |                           |                    |                    |                    |                    |
```

File: proposals/simd/ImplementationStatus.md [hunk 3]
```diff
@@ -174,6 +178,10 @@
 | `f64x2.div`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
 | `f64x2.min`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
 | `f64x2.max`                |               `-msimd128` | :heavy_check_mark: |                    |                    | :heavy_check_mark: |
+| `f64x2.nearest`            |                           |                    |                    |                    |                    |
+| `f64x2.trunc`              |                           |                    |                    |                    |                    |
+| `f64x2.ceil`               |                           |                    |                    |                    |                    |
+| `f64x2.floor`              |                           |                    |                    |                    |                    |
```
Comment (tlively) [inline proposals/simd/ImplementationStatus.md:184]:
```suggestion
| `f64x2.ceil`               |                           |                    |                    |                    |                    |
| `f64x2.floor`              |                           |                    |                    |                    |                    |
| `f64x2.trunc`              |                           |                    |                    |                    |                    |
| `f64x2.nearest`            |                           |                    |                    |                    |                    |
```

File: proposals/simd/SIMD.md [hunk 4]
```diff
@@ -900,6 +900,30 @@ Lane-wise IEEE `multiplication`.
...
+* `f64x2.ceil(a: v128) -> v128`
+
+Lane-wise rounding to the nearest integral value not smaller than the input.
+
+### Round to integer below (floor)
+* `f32x4.floor(a: v128) -> v128`
+* `f64x2.floor(a: v128) -> v128`
+
+Lane-wise rounding to the nearest integral value not greater than the input.
+
```
Comment (tlively) [inline proposals/simd/SIMD.md:926]:
```suggestion
### Round to integer above (ceiling)
* `f32x4.ceil(a: v128) -> v128`
* `f64x2.ceil(a: v128) -> v128`

Lane-wise rounding to the nearest integral value not smaller than the input.

### Round to integer toward zero (truncate to integer)
* `f32x4.trunc(a: v128) -> v128`
* `f64x2.trunc(a: v128) -> v128`

Lane-wise rounding to the nearest integral value with the magnitude not larger than the input.

### Round to integer below (floor)
* `f32x4.floor(a: v128) -> v128`
* `f64x2.floor(a: v128) -> v128`

Lane-wise rounding to the nearest integral value not greater than the input.

### Round to nearest integer, ties to even
* `f32x4.nearest(a: v128) -> v128`
* `f64x2.nearest(a: v128) -> v128`

Lane-wise rounding to the nearest integral value; if two values are equally near, rounds to the even one.

```

**Comment by dtig** [comment]
Thanks @Maratyszcza for filing the issues, moving this to prototyping as on all platforms that we are using as a baseline currently these have a direct mapping to instructions, and on ARMv7, there is a precedent for them being slow as this is the case for the scalar versions of these operations as well, some implementations call out to the runtime to implement them. Moving to pending prototype data as we are prototyping them in V8, adding a retroactive label update.

**Comment by tlively** [comment]
These will be available in the next version of Emscripten via `__builtin_wasm_{ceil,floor,trunc,nearest}_{f32x4,f64x2}`.

**Comment by ngzhian** [comment]
Prototype in V8 is done for x64, ia32, ARM64. Still working on ARM.

**Comment by Maratyszcza** [comment]
I evaluated these instructions on the rounding operators in the [XNNPACK library](https://github.com/google/XNNPACK), and the results are below:

Operator | Speedup
-- | --
Bankers' Rounding | 2.01X
Ceiling | 2.88X
Floor | 2.45X
Truncation | 2.05X

These speedups were measured on ThinkStation P520, and the code modifications can be seen in google/XNNPACK#782

**Comment by ngzhian** [comment]
This has been accepted into the proposal [0] during the sync on 2020-09-04. This LGTM, as it is.

Note, I would like https://github.com/WebAssembly/simd/blob/master/proposals/simd/NewOpcodes.md to be updated too, but it requires more tweaks (since there is a bit of a collision in opcodes for these instructions and the "reserved ones" under i64x2, and also ordering of instructions for presentation). But that's not a big problem, and can be worked on in the future.

[0] https://docs.google.com/document/d/138cF6aOUa9RZC2tOR7AhlIQWdmX5EtpzXRTVDAN3bfo/edit# see "4. Floating point rounding"

File: proposals/simd/SIMD.md [hunk 5]
```diff
@@ -900,6 +900,30 @@ Lane-wise IEEE `multiplication`.
 
 Lane-wise IEEE `squareRoot`.
 
+### Round to integer above (ceiling)
```
Comment (ngzhian) [inline proposals/simd/SIMD.md:903]: Nit: can you reorder this? We have ceil, floor, trunc, nearest in the instruction, this section is ceil, trunc, floor, nearest.

Comment (Maratyszcza) [reply proposals/simd/SIMD.md:903]: Done

**Comment by ngzhian** [comment]
@tlively this wasn't added to NewOpcodes.md, just fyi in case you are looking at that doc for opcode organization.

**Comment by tlively** [comment]
Oh, thanks for point that out. I had indeed missed them.

**Pull request merged**